File: number-theoretic-transform-integer-dft/numbertheoretictransform.py
# 
# Number-theoretic transform library (Python)
# 
# Copyright (c) 2022 Project Nayuki
# All rights reserved. Contact Nayuki for licensing.
# https://www.nayuki.io/page/number-theoretic-transform-integer-dft
# 
import itertools
from typing import List, Tuple


# ---- High-level NTT functions ----

def find_params_and_transform(invec: List[int], minmod: int) -> Tuple[List[int], int, int]:
    mod: int = find_modulus(len(invec), minmod)
    root: int = find_primitive_root(len(invec), mod - 1, mod)
    return (transform(invec, root, mod), root, mod)


def transform(invec: List[int], root: int, mod: int) -> List[int]:
    if len(invec) >= mod:
        raise ValueError()
    if not all((0 <= val < mod) for val in invec):
        raise ValueError()
    if not (1 <= root < mod):
        raise ValueError()

    outvec: List[int] = []
    for i in range(len(invec)):
        temp: int = 0
        for (j, val) in enumerate(invec):
            temp += val * pow(root, i * j, mod)
            temp %= mod
        outvec.append(temp)
    return outvec


def inverse_transform(invec: List[int], root: int, mod: int) -> List[int]:
    outvec: List[int] = transform(invec, pow(root, -1, mod), mod)
    scaler: int = pow(len(invec), -1, mod)
    return [(val * scaler % mod) for val in outvec]


def transform_radix_2(vector: List[int], root: int, mod: int) -> None:
    n: int = len(vector)
    levels: int = n.bit_length() - 1
    if 1 << levels != n:
        raise ValueError("Length is not a power of 2")

    def reverse(x: int, bits: int) -> int:
        return int(format(x, f'0{bits}b')[::-1], 2)

    for i in range(n):
        j: int = reverse(i, levels)
        if j > i:
            vector[i], vector[j] = vector[j], vector[i]

    powtable: List[int] = []
    temp: int = 1
    for i in range(n // 2):
        powtable.append(temp)
        temp = (temp * root) % mod

    size: int = 2
    while size <= n:
        halfsize: int = size >> 1
        tablestep: int = n // size
        for i in range(0, n, size):
            k: int = 0
            for j in range(i, i + halfsize):
                l: int = j + halfsize
                left: int = vector[j]
                right: int = vector[l] * powtable[k] % mod
                vector[j] = (left + right) % mod
                vector[l] = (left - right) % mod
                k += tablestep
        size <<= 1


def circular_convolve(vec0: List[int], vec1: List[int]) -> List[int]:
    if not (0 < len(vec0) == len(vec1)):
        raise ValueError()
    if any((val < 0) for val in itertools.chain(vec0, vec1)):
        raise ValueError()
    maxval: int = max(val for val in itertools.chain(vec0, vec1))
    minmod: int = maxval ** 2 * len(vec0) + 1
    temp0, root, mod = find_params_and_transform(vec0, minmod)
    temp1: List[int] = transform(vec1, root, mod)
    temp2: List[int] = [(x * y % mod) for (x, y) in zip(temp0, temp1)]
    return inverse_transform(temp2, root, mod)


# ---- Mid-level number theory functions for NTT ----

def find_modulus(veclen: int, minimum: int) -> int:
    if veclen < 1 or minimum < 1:
        raise ValueError()
    start: int = (minimum - 1 + veclen - 1) // veclen
    for i in itertools.count(max(start, 1)):
        n: int = i * veclen + 1
        assert n >= minimum
        if is_prime(n):
            return n
    raise AssertionError("Unreachable")


def find_primitive_root(degree: int, totient: int, mod: int) -> int:
    if not (1 <= degree <= totient < mod):
        raise ValueError()
    if totient % degree != 0:
        raise ValueError()
    gen: int = find_generator(totient, mod)
    root: int = pow(gen, totient // degree, mod)
    assert 0 <= root < mod
    return root


def find_generator(totient: int, mod: int) -> int:
    if not (1 <= totient < mod):
        raise ValueError()
    for i in range(1, mod):
        if is_primitive_root(i, totient, mod):
            return i
    raise ValueError("No generator exists")


def is_primitive_root(val: int, degree: int, mod: int) -> bool:
    if not (0 <= val < mod):
        raise ValueError()
    if not (1 <= degree < mod):
        raise ValueError()
    pf: List[int] = unique_prime_factors(degree)
    return pow(val, degree, mod) == 1 and \
           all((pow(val, degree // p, mod) != 1) for p in pf)


# ---- Low-level common number theory functions ----

def unique_prime_factors(n: int) -> List[int]:
    if n < 1:
        raise ValueError()
    result: List[int] = []
    i: int = 2
    end: int = sqrt(n)
    while i <= end:
        if n % i == 0:
            n //= i
            result.append(i)
            while n % i == 0:
                n //= i
            end = sqrt(n)
        i += 1
    if n > 1:
        result.append(n)
    return result


def is_prime(n: int) -> bool:
    if n <= 1:
        raise ValueError()
    return all((n % i != 0) for i in range(2, sqrt(n) + 1))


def sqrt(n: int) -> int:
    if n < 0:
        raise ValueError()
    i: int = 1
    while i * i <= n:
        i <<= 1
    result: int = 0
    while i > 0:
        if (result + i) ** 2 <= n:
            result += i
        i >>= 1
    return result

File: number-theoretic-transform-integer-dft/numbertheoretictransform.js
/*
 * Number-theoretic transform library (compiled from TypeScript)
 *
 * Copyright (c) 2022 Project Nayuki
 * All rights reserved. Contact Nayuki for licensing.
 * https://www.nayuki.io/page/number-theoretic-transform-integer-dft
 */
"use strict";
var numbertheoretictransform;
(function (numbertheoretictransform) {
    /*---- High-level NTT functions ----*/
    // Returns the forward number-theoretic transform of the given vector with
    // respect to the given primitive nth root of unity under the given modulus.
    function transform(invec, root, mod) {
        const n = invec.length;
        let outvec = [];
        for (let i = 0; i < n; i++) {
            let sum = 0n;
            for (let j = 0; j < n; j++) {
                const k = BigInt(i * j % n);
                sum = (sum + invec[j] * powMod(root, k, mod)) % mod;
            }
            outvec.push(sum);
        }
        return outvec;
    }
    numbertheoretictransform.transform = transform;
    // Returns the inverse number-theoretic transform of the given vector with
    // respect to the given primitive nth root of unity under the given modulus.
    function inverseTransform(invec, root, mod) {
        let outvec = transform(invec, reciprocalMod(root, mod), mod);
        let scaler = reciprocalMod(BigInt(invec.length), mod);
        for (let i = 0; i < outvec.length; i++)
            outvec[i] = (outvec[i] * scaler) % mod;
        return outvec;
    }
    numbertheoretictransform.inverseTransform = inverseTransform;
    // Computes the forward number-theoretic transform of the given vector in place,
    // with respect to the given primitive nth root of unity under the given modulus.
    // The length of the vector must be a power of 2.
    function transformRadix2(vector, root, mod) {
        let n = vector.length;
        let levels = n.toString(2).length - 1;
        if (1 << levels != n)
            throw new RangeError("Length is not a power of 2");
        for (let i = 0; i < n; i++) {
            const j = reverseBits(i, levels);
            if (j > i) {
                const temp = vector[i];
                vector[i] = vector[j];
                vector[j] = temp;
            }
        }
        let powTable = new Array(Math.floor(n / 2));
        {
            let temp = 1n;
            for (let i = 0; i < powTable.length; i++) {
                powTable[i] = temp;
                temp = temp * root % mod;
            }
        }
        for (let size = 2; size <= n; size *= 2) {
            const halfsize = size / 2;
            const tablestep = n / size;
            for (let i = 0; i < n; i += size) {
                for (let j = i, k = 0; j < i + halfsize; j++, k += tablestep) {
                    let l = j + halfsize;
                    const left = vector[j];
                    const right = vector[j + halfsize] * powTable[k];
                    vector[j] = (left + right) % mod;
                    vector[l] = (left - right) % mod;
                }
            }
            if (size == n)
                break;
        }
        // Returns the integer whose value is the reverse of the lowest 'width' bits of the integer 'val'.
        function reverseBits(val, width) {
            let result = 0;
            for (let i = 0; i < width; i++) {
                result = (result << 1) | (val & 1);
                val >>>= 1;
            }
            return result;
        }
    }
    numbertheoretictransform.transformRadix2 = transformRadix2;
    // Returns the circular convolution of the given vectors of integers.
    // All values must be non-negative. Internally, a sufficiently large modulus
    // is chosen so that the convolved result can be represented without overflow.
    function circularConvolve(vec0, vec1) {
        if (vec0.length == 0 || vec0.length != vec1.length)
            throw new RangeError();
        let maxval = 0n;
        for (const x of vec0.concat(vec1)) {
            if (x > maxval)
                maxval = x;
        }
        const minmod = maxval * maxval * BigInt(vec0.length) + 1n;
        const mod = findModulus(vec0.length, minmod);
        const root = findPrimitiveRoot(BigInt(vec0.length), mod - 1n, mod);
        const temp0 = transform(vec0, root, mod);
        const temp1 = transform(vec1, root, mod);
        let temp2 = [];
        for (let i = 0; i < temp0.length; i++)
            temp2.push(temp0[i] * temp1[i] % mod);
        return inverseTransform(temp2, root, mod);
    }
    numbertheoretictransform.circularConvolve = circularConvolve;
    /*---- Mid-level number theory functions for NTT ----*/
    // Returns the smallest modulus mod such that mod = i * veclen + 1
    // for some integer i >= 1, mod > veclen, and mod is prime.
    // Although the loop might run for a long time and create arbitrarily large numbers,
    // Dirichlet's theorem guarantees that such a prime number must exist.
    function findModulus(vecLen, minimum) {
        if (vecLen < 1 || minimum < 1n)
            throw new RangeError();
        const vl = BigInt(vecLen);
        let start = (minimum - 1n + vl - 1n) / vl;
        if (start < 1n)
            start = 1n;
        for (let n = start * vl + 1n;; n += vl) {
            if (isPrime(n))
                return n;
        }
    }
    numbertheoretictransform.findModulus = findModulus;
    // Returns an arbitrary primitive degree-th root of unity modulo mod.
    // totient must be a multiple of degree. If mod is prime, an answer must exist.
    function findPrimitiveRoot(degree, totient, mod) {
        if (!(0 <= degree && degree <= totient && totient < mod))
            throw new RangeError();
        if (totient % degree != 0n)
            throw new RangeError();
        const gen = findGenerator(totient, mod);
        return powMod(gen, totient / degree, mod);
    }
    numbertheoretictransform.findPrimitiveRoot = findPrimitiveRoot;
    // Returns an arbitrary generator of the multiplicative group of integers modulo mod.
    // totient must equal the Euler phi function of mod. If mod is prime, an answer must exist.
    function findGenerator(totient, mod) {
        if (!(1n <= totient && totient < mod))
            throw new RangeError();
        for (let i = 1n; i < mod; i++) {
            if (isPrimitiveRoot(i, totient, mod))
                return i;
        }
        throw new Error("No generator exists");
    }
    numbertheoretictransform.findGenerator = findGenerator;
    // Tests whether val is a primitive degree-th root of unity modulo mod.
    // In other words, val^degree % mod = 1, and for each 1 <= k < degree, val^k % mod != 1.
    // 
    // To test whether val generates the multiplicative group of integers modulo mod,
    // set degree = totient(mod), where totient is the Euler phi function.
    // We say that val is a generator modulo mod if and only if the set of numbers
    // {val^0 % mod, val^1 % mod, ..., val^(totient-1) % mod} is equal to the set of all
    // numbers in the range [0, mod) that are coprime to mod. If mod is prime, then
    // totient = mod - 1, and powers of a generator produces all integers in the range [1, mod).
    function isPrimitiveRoot(val, degree, mod) {
        if (!(0n <= val && val < mod))
            throw new RangeError();
        if (!(1n <= degree && degree < mod))
            throw new RangeError();
        return powMod(val, degree, mod) == 1n &&
            uniquePrimeFactors(degree).every(p => powMod(val, degree / p, mod) != 1n);
    }
    numbertheoretictransform.isPrimitiveRoot = isPrimitiveRoot;
    /*---- Low-level common number theory functions ----*/
    // Returns a list of unique prime factors of the given integer in
    // ascending order. For example, uniquePrimeFactors(60) = [2, 3, 5].
    function uniquePrimeFactors(n) {
        if (n < 1n)
            throw new RangeError();
        let result = [];
        for (let i = 2n, end = sqrt(n); i <= end; i++) {
            if (n % i == 0n) {
                result.push(i);
                do
                    n = n / i;
                while (n % i == 0n);
                end = sqrt(n);
            }
        }
        if (n > 1n)
            result.push(n);
        return result;
    }
    numbertheoretictransform.uniquePrimeFactors = uniquePrimeFactors;
    // Tests whether the given integer n >= 2 is a prime number.
    function isPrime(n) {
        if (n <= 1n)
            throw new RangeError();
        if ((n & 1n) == 0n)
            return n == 2n;
        for (let i = 3n, end = sqrt(n); i <= end; i += 2n) {
            if (n % i == 0n)
                return false;
        }
        return true;
    }
    numbertheoretictransform.isPrime = isPrime;
    // Returns floor(sqrt(x)) for the given integer x >= 0.
    function sqrt(x) {
        if (x < 0n)
            throw new RangeError();
        let y = 0n;
        for (let i = BigInt(Math.floor((x.toString(2).length - 1) / 2)); i >= 0n; i--) {
            y |= 1n << i;
            if (y * y > x)
                y ^= 1n << i;
        }
        return y;
    }
    numbertheoretictransform.sqrt = sqrt;
    // Returns x^y mod m.
    function powMod(x, y, mod) {
        if (y < 0n || mod <= 0n)
            throw RangeError();
        if (!(0n <= x && x < mod))
            x = ((x % mod) + mod) % mod;
        let result = 1n;
        while (y != 0n) {
            if ((y & 1n) != 0n)
                result = result * x % mod;
            x = x * x % mod;
            y >>= 1n;
        }
        return result;
    }
    numbertheoretictransform.powMod = powMod;
    // Returns x^-1 mod m.
    function reciprocalMod(x, mod) {
        if (!(0n <= x && x < mod))
            throw RangeError();
        // Based on a simplification of the extended Euclidean algorithm
        let y = x;
        x = mod;
        let a = 0n;
        let b = 1n;
        while (y != 0n) {
            let temp = a - x / y * b;
            a = b;
            b = temp;
            temp = x % y;
            x = y;
            y = temp;
        }
        if (x == 1n)
            return ((a % mod) + mod) % mod;
        else
            throw new RangeError("Reciprocal does not exist");
    }
    numbertheoretictransform.reciprocalMod = reciprocalMod;
})(numbertheoretictransform || (numbertheoretictransform = {}));
